designs/v1beta1-full-changelist.md

```diff
@@ -0,0 +1,200 @@
+    1. To support *some* of these use-cases, Karpenter has begun effectively creating “system-tags” i.e. (`aws::ids`, `aws::owners`, `aws::name`). These are special-cased version of the standard user custom-tags that allow users to achieve the scenarios described in #2; however, they are not easily discoverable or understood and if we are beginning to support special-cases like this, it makes sense that we should begin to structure these fields.
+
+```
+amiSelector:
+- name: foo
+  id: abc-123
+  owner: amazon
+  ssm: "/al2/alias"
+  tags: 
+    key: value
+# Selectors are ORed
+- name: foo
+  id: abc-123
+  owner: self
+  ssm: "/al2/alias"
+  tags: 
+    key: value
+```
+
+#### Update `nodeClass.spec.subnetSelector`
+
+`subnetSelector` should have a similar parity to the `amiSelector` in its design to improve the ease-of-use for users. As a result, we should design the `subnetSelector` in the same spirit as the `amiSelector` such that you can also specify multiple selectors through `tags` and `ids` that can be ORed together to produce the ultimate set of items that you want to use.
+
+```
+subnetSelector:
+- id: abc-123
+  tags:
+    key: value
+# Selectors are ORed
+- id: abc-123
+  tags:
+    key: value
+```
+
+#### Update `nodeClass.spec.securityGroupSelector`
+
+The same logic for `subnetSelector` applies to `securityGroupSelector`. We should have a similar parity to the `amiSelector` to improve the ease-of-use around this selector.
+
+```
+securityGroupSelector:
+- id: abc-123
+  name: default-secuirty-group # Not the same as the name tag
+  tags:
+    key: value
+# Selectors are ORed
+- id: abc-123
+  name: custom-security-group # Not the same as the name tag
+  tags:
+    key: value
+```
+
+#### Remove `nodeClass.spec.launchTemplate`
+
+Direct launch template support is problematic for many reasons, outlined in the design [Unmanaged LaunchTemplate Support for Karpenter](./unmanaged-launch-template-removal.md). Customers continue to run into issues when directly using launch templates. Rather than continue to maintain these sharp edges and give users a half-baked experience of Karpenter, we should remove this field, considering that we can always add it back later if there is enough ask from users to do so.
+
+#### `nodeTemplate.spec.instanceProfile` → `nodeClass.spec.role`
+
+Currently, Karpenter uses an `instanceProfile` in the `AWSNodeTemplate` that is referenced to determine the profile that the EC2 node should launch with. Instance profiles are IAM entities that are specific to EC2 and do not have a lot of detail built around them (including console support); users are generally more familiar with the concept of IAM roles. As a result, we can support a `role` in the new `NodeClass` and allow Karpenter to provision the instance profile `ad-hoc` with the `role`  specified attached to it.
```

Comment:
If I want to use an instance profile and _not_ allow Karpenter to manage that, can I still have that option?

Reply:
What would be the reason that you would want to use your own instance profile. I think we can add this back if we hear enough asks for it but there doesn't seem to be a strong enough reason to keep it when there is a 1-1 mapping between a role and an instance profile

Reply:
I might want to use a role that some other code has made, along with an instance profile.

Reply:
Still not sure that I get the full use-case. It doesn't seem likely to me that you would have a hard requirement to utilize a specific instance profile that must be manually specified. To me, it seems like you would enforce this restriction at the role level.

Reply:
If the instance profile already exists, what behavior / error do I see?
If I stop using that role, does Karpenter try to delete the instance profile that is no longer needed?

Reply:
Karpenter will surface the instance profile through a discovery mechanism and resolve this instance profile into the AWS `NodeClass` status. If the role changes, Karpenter should go and cleanup the old role assignment and update it with the new one. 

We should be generating the instanceProfile with a similar mechanism to `generateName` so that it is sufficiently random that we are unlikely to have collisions; however, if we do have a collision, we should just try to regenerate a random value again.

Reply:
OK, cool. There need not be a 1-to-1 mapping though. _n_ instance profiles can map to 1 role (right?)

Given that Karpenter is going to manage instance profiles, I might want:
- to specify a custom `path` for the instance profile
- to set custom tags on the instance profile

More likely: I restrict the IAM role that Karpenter runs as. The restrictions enforce tag-on-create for instance profiles, and expect certain tags to be set. Karpenter is then only allowed to delete or update instance profiles that have the appropriate tag.